File: mobile/flutter/ascend_flutter/lib/features/posts/data/datasources/posts_remote_data_source.dart
import 'package:ascend_flutter/core/network/api_constants.dart';
import 'package:dio/dio.dart';

import '../models/post_model.dart';

abstract class PostsRemoteDataSource {
  Future<List<PostModel>> getPosts();
}

class PostsRemoteDataSourceImpl implements PostsRemoteDataSource {
  final Dio dio;

  const PostsRemoteDataSourceImpl(this.dio);

  @override
  Future<List<PostModel>> getPosts() async {
    final response = await dio.get(ApiConstants.posts);

    final data = response.data as List;

    return data
        .map((json) => PostModel.fromJson(json as Map<String, dynamic>))
        .toList();
  }
}

File: mobile/react-native/ascend_rn/src/features/posts/data/datasources/PostsRemoteDataSource.ts
import { apiClient } from '../../../../core/network/apiClient';
import { ApiConstants } from '../../../../core/network/apiConstants';

import { PostModel } from '../models/PostModel';

export interface PostsRemoteDataSource {
  getPosts(): Promise<PostModel[]>;
}

export class PostsRemoteDataSourceImpl
  implements PostsRemoteDataSource {

  async getPosts(): Promise<PostModel[]> {
    const response = await apiClient.get(
      ApiConstants.posts,
    );

    return response.data.map(
      (json: any) => PostModel.fromJson(json),
    );
  }
}